### ISSUE TITLE
HTTP Metricbeat module fails requesting paths that contain query parameters

### ISSUE THREAD

**Issue by sayden** [elastic/beats#8286]
TL;DR Configuring a `path` in `http` metricbeat module that contains something like `/mypath?key=value` makes the module do not reach the proper server path

Following this Discuss issue https://discuss.elastic.co/t/http-module-invalid-character-looking-for-beginning-of-value/148082 it seems that metricbeat can't reach the proper server path when configuring `http` module with a `path` like the following:

```yaml
metricbeat.modules:
- module: http
  metricsets:
   - json
   - server
  period: 1s
  hosts: ["localhost:8080"]
  namespace: "json_namespace"
  path: "/home?key=value"
output.console:
  pretty: true
```

This can be check with a dummy server like the following:
```go
package main

import (
	"net/http"
	"fmt"
	"github.com/thehivecorporation/log"
)

func main() {
	http.HandleFunc("/home", func(w http.ResponseWriter, r *http.Request){
		values := r.URL.Query()
		value := values.Get("key")

		log.WithField("key", value).Info("Request received")

		w.Write([]byte(fmt.Sprintf(`{"hello":"world", "key":"%s"}`, value)))
	})

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request){
		log.Warn("Request lost")
		w.WriteHeader(404)
	})

	http.ListenAndServe(":8080", nil)
}
```
Which produces the following output in the server console:
```shell
WARN[0005](2018-09-11T15:52:35.071654578+02:00) github.com/sayden/dummyserver/main.go:20 Request lost
```

In this particular case, the expected behaviour is that the first `HandleFunc` handles the request to the server and producing an output like the following:
```json
{
  "@timestamp": "2018-09-12T08:48:11.127Z",
  "@metadata": {
    "beat": "metricbeat",
    "type": "doc",
    "version": "6.4.0"
  },
  "host": {
    "name": "caster-apricity"
  },
  "metricset": {
    "rtt": 6087,
    "namespace": "http.json_namespace",
    "name": "json",
    "module": "http",
    "host": "localhost:8080"
  },
  "http": {
    "json_namespace": {
      "hello": "world",
      "key": "value"
    }
  },
  "beat": {
    "version": "6.4.0",
    "name": "caster-apricity",
    "hostname": "caster-apricity"
  }
}
```

Instead, you receive something like this:
```json
{
  "@timestamp": "2018-09-12T08:45:25.277Z",
  "@metadata": {
    "beat": "metricbeat",
    "type": "doc",
    "version": "6.4.0"
  },
  "host": {
    "name": "caster-apricity"
  },
  "metricset": {
    "name": "json",
    "module": "http",
    "host": "localhost:8080",
    "rtt": 616
  },
  "error": {
    "message": "unexpected end of JSON input"
  },
  "beat": {
    "name": "caster-apricity",
    "hostname": "caster-apricity",
    "version": "6.4.0"
  }
}
```

**Comment by exekias** [comment]
Thanks for opening this @sayden! I think the expected output is missing?

Also, I'm guessing that this bug affects all modules using the [HTTP helper](https://github.com/elastic/beats/blob/master/metricbeat/helper/http.go)

**Comment by sayden** [comment]
Thanks for the thumbs up @exekias I've updated the comment to show the expected and current output.

I'll also take a look at the HTTP Helper, thanks again for the thumbs up too. :)

**Comment by sayden** [comment]
Ok, I found that the root of the issue in within `net/url` package. When calling `String() string` method on a `url.URL` it will call `EscapedPath` as you can see here https://github.com/golang/go/blob/master/src/net/url/url.go#L797

The thing is that I feel like this is a correct behaviour because the `net/url` package has its own mechanisms to deal with query params https://github.com/golang/go/blob/master/src/net/url/url.go#L872

We actually generate our `*url.Url` properly using the all available data we have here https://github.com/elastic/beats/blob/master/metricbeat/mb/parse/url.go#L170 (doing a unit test with this function and passing query params on the `query` param of the functions solves the problem).

I'm considering two options:
1. Improve the base parser with a new query configuration field. This will fix the problem in any module using `mb/parse` package
2. Add the query configuration field only on the `http` module. I guess a safer solution because less pieces will be involved but we won't solve the root of the problem for the rest of the modules.

The thing is that I need some guidance to understand how the configuration is set in this case
1. Here https://github.com/elastic/beats/blob/master/metricbeat/mb/mb.go#L301 we could add the new field to the entire `mb` but then it must be promoted to use within any module that uses the package. I'm also not sure of how this will work with a customized *hostparser* like the one you can see in `http/json` metricset (see next link to see the hostparser).
2. Here https://github.com/elastic/beats/blob/master/metricbeat/module/http/json/json.go#L49 the host parser of the `http` module and `json` metricset (the one that the user was using in the issue) is set with default attributes but we'll need to set it with whatever comes from the *new* configuration field.

Ideas? :)

**Comment by exekias** [comment]
How about updating the HTTP helper itself? We recently did some changes to it to support `headers` for instance: https://github.com/elastic/beats/pull/3945

Something that passes these params as part of the request should be possible there, and would fix the issue for all modules relying on the helper.

**Comment by sayden** [comment]
Ok, I have uploaded a PR that corrects the issue in the `mb/parse` package while I keep digging into a proper way to implement it in the HTTP helper.

The thing is that the helper doesn't have access to the Parser so any implementation within the HTTP helper will generate unpredictable code.

You can take a look here to continue the discussion about the proper place to implement this: https://github.com/elastic/beats/pull/8292

**Comment by sayden** [comment]
Fixed in #8292
